internal/throttling/memory_gcra.go

```diff
@@ -0,0 +1,63 @@
+package throttling
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/throttled/throttled/v2"
+	"github.com/throttled/throttled/v2/store/memstore"
+
+	"github.com/rudderlabs/rudder-server/internal/throttling/cachettl"
+)
+
+const defaultMaxCASAttemptsLimit = 100
+
+type gcra struct {
+	mu    sync.Mutex
+	store *cachettl.Cache[string, *throttled.GCRARateLimiter]
+}
+
+func (g *gcra) limit(key string, cost, burst, rate, period int64) (
+	bool, error,
+) {
+	rl, err := g.getLimiter(key, burst, rate, period)
+	if err != nil {
+		return false, err
+	}
+
+	limited, _, err := rl.RateLimit("key", int(cost))
+	if err != nil {
+		return false, fmt.Errorf("could not rate limit: %w", err)
+	}
+
+	return !limited, nil
+}
+
+func (g *gcra) getLimiter(key string, burst, rate, period int64) (*throttled.GCRARateLimiter, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.store == nil {
+		g.store = cachettl.New[string, *throttled.GCRARateLimiter]()
+	}
+
+	rl := g.store.Get(key)
+	if rl == nil {
+		store, err := memstore.New(0)
```

Comment:
You're gonna need an astonishing number of destinations to be able to blow this up. It's true that the throttler can be used in other places too but I don't think setting a limit here makes much sense anyway. If `0` it just means that internally it uses a `map[string]*int64` instead of a LRU data structure with a limit. Like in other places where maps and other data structures that can grow are used, I think it's sensible to pay attention on the caller side instead of setting a hard limit in the data structure itself (e.g. instead of setting a limit on how many keys a map should have it's more sensible to avoid things like loading all data from a database into a map).